homeassistant/components/logbook/__init__.py

```diff
@@ -435,79 +435,116 @@ def _get_events(
 
     def yield_rows(query: Query) -> Generator[Row, None, None]:
         """Yield Events that are not filtered away."""
-        for row in query.yield_per(1000):
+        if entity_ids or context_id:
+            rows = query.all()
+        else:
+            rows = query.yield_per(1000)
+        for row in rows:
             context_lookup.setdefault(row.context_id, row)
-            if row.event_type != EVENT_CALL_SERVICE and (
-                row.event_type == EVENT_STATE_CHANGED
-                or _keep_row(hass, row, entities_filter)
+            event_type = row.event_type
+            if event_type != EVENT_CALL_SERVICE and (
```

Comment:
Can we filter `call_service` out with the query instead of in Python?

Reply:
Sadly no. We need it so we can track it's context id for when it's the source of the change for another event


With the current implementation we are actually selecting all the relevant events during the period even when we specify entities because we need the context ids. I am hoping to optimize that in another turn but haven't come up with a working strategy yet.

Reply:
Solved here https://github.com/home-assistant/core/pull/71657